src/System.Linq.Expressions/src/System/Linq/Expressions/StackGuard.cs

```diff
@@ -16,16 +16,7 @@ internal sealed class StackGuard
 
         public bool TryEnterOnCurrentStack()
         {
-            try
-            {
-                RuntimeHelpers.EnsureSufficientExecutionStack();
-            }
-            catch (InsufficientExecutionStackException) when (_executionStackCount < MaxExecutionStackCount)
-            {
-                return false;
-            }
-
-            return true;
+            return RuntimeHelpers.TryEnsureSufficientExecutionStack() || _executionStackCount >= MaxExecutionStackCount;
```

Comment:
Let's stick with the exception for now and use this PR just to switch from the exception to the Try.  There's a big difference between throwing an exception and crashing.